### PR TITLE
Datatype

### PR DESCRIPTION
**Pull request by xnning**
nested pattern match.
e.g. 
data MList = MNil | MCons Int MList;
let f (x:MList) =
  case x of MNil -> 0
            | (MCons x (MCons y _)) -> x + y
            | _ -> 2

constructor can be partially applied.
e.g.

let g (x:MList -> MList) = x MNil;
g (MCons 3)

This is a primitive version. May need further optimization.

### CODE REVIEW

**Comment by tomtau** [comment]
are there any tests that use datatype in the test suite?

File: lib/ClosureF.hs [hunk 1]
```diff
@@ -66,17 +66,19 @@ data Expr t e =
...
    | JMethod (Either ClassName (Expr t e)) MethodName [Expr t e] ClassName
-   | JField  (Either ClassName (Expr t e)) FieldName ClassName
+   | JField  (Either ClassName (Expr t e)) FieldName (Type t)
    | PolyList [Expr t e] (Type t)
    | JProxyCall (Expr t e) (Type t)
    | SeqExprs [Expr t e]
    | Data S.RecFlag [DataBind t] (Expr t e)
    | Constr (Constructor t) [Expr t e]
    | Case (Expr t e) [Alt t e]
+   | CaseCast (Expr t e) (Constructor t)
```
Comment (tomtau) [inline lib/ClosureF.hs]: what's this for?

Comment (xnning) [inline lib/ClosureF.hs]: While entering a pattern, cast is needed.
Like

```
case x of
    Cons  y z -> case y of ...
```

Here `y` will become `JField (CaseCast x "Cons") "field1"`, namely `(Cons)x.field1` in java

I have thought about not adding `CaseCast` but have not figured out a better alternative way. Any suggestions?

**Comment by xnning** [comment]
@tomtau  No. There are a few examples living in the examples folder.

**Comment by tomtau** [comment]
@xnning currently, the backend will be rewritten - in order to be sure that one does not break these new cases when rewriting it, there should be some tests in the test suite. Please add some - I think the standard test suite is still in migration ( @zhiyuanshi ?), so you can use `make test2` instead

**Comment by zhiyuanshi** [comment]
@tomtau Right. `make test2` is the standard right now.

File: lib/ClosureF.hs (hunk 1 above)
Comment (tomtau) [inline lib/ClosureF.hs]: Using reflection methods:
`Class.forName("Cons").cast(x.field1);`

Comment (xnning) [inline lib/ClosureF.hs]: Hmm, don't quiet understand. How to represent the cast behavior in ClosureF without adding a Expr?
Btw, seems i made it confused. It is `((Cons)x).field1` .

Comment (tomtau) [inline lib/ClosureF.hs]: then it's 
`Class.forName("Cons").cast(x).field1`

Not sure if it's possible to call static methods though - @lihuanglx ?

Comment (xnning) [inline lib/ClosureF.hs]: ooh, I see! I will try it later.
Yes it can call static methods

File: lib/BaseTransCFJava.hs [hunk 2]
```diff
@@ -411,63 +415,46 @@ trans self =
...
-                do let Datatype nam _ _ = last ctrParams
-                       ctrName = nam ++ ctrName'
-                   es' <- mapM (translateM this) es
-                   newVarName <- getNewVarName this
-                   let (stmts, oexpr, _) =  concatFirst $ unzip3 es'
-                   let inst = case es' of
-                                [] -> localVar (classTy nam) (varDecl newVarName $ methodCallExp [map toLower nam,ctrName] [])
-                                _  -> localVar (classTy nam) (varDecl newVarName $ methodCallExp [map toLower nam,ctrName] (map unwrap oexpr))
-                   return (stmts ++ [inst], var newVarName, last ctrParams)
+              Constr (Constructor ctrName' ctrParams) es -> do
```
Comment (tomtau) [inline lib/BaseTransCFJava.hs:428]: what's FVar?

Comment (xnning) [inline lib/BaseTransCFJava.hs:428]: don't know about it. maybe @bixuanzju knows?

Comment (bixuanzju) [inline lib/BaseTransCFJava.hs:428]: No idea either, i will run `git blame` to find who added that ...

Comment (bixuanzju) [inline lib/BaseTransCFJava.hs:428]: @zhiyuanshi https://github.com/hkuplg/fcore/commit/5bb665c2693a0d4aac0a2a34198296c2411bc6bc#diff-beea0fb4f2da1c14085d156e07fd5b49R28

**Pull request merged**